File: src/Controls/KarveControls/DataGrid/DataGridHelpers/DataGridColumns/DataGridBrowseBoxColumn.cs

using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Drawing;
using System.Windows.Media;
using Telerik.WinControls.UI;

public class DataGridBrowseBoxColumn : GridViewBrowseColumn
{

    #region "Variables"
    protected int _Item;
    protected string _Campo;
    protected string _Tabla;
    protected string _ExpresionBd;
    protected GridViewDataColumn _relatedColumn;

   // protected System.Drawing.Color _BackColor = Drawing.Color.White;
    private string _Desc_Datafield;
    //tabla con la que se rellena el campo descripcion
    private string _Desc_DBTable;
    //referencia de la tabla descripcion (el textbox editable)
    private string _Desc_DBPK;
    //filtro extra en la consulta
    private string myWhere;
    //si tiene que volver a tirar la query (se pone en true cuando el texto cambia)
    private bool _requery = true;
    //si tira la query cada vez que el texto cambia o espera al lost focus, si por defecto
    private bool _queryOnTextChanged = true;
    protected int _txtDataWidth;
    private string _extraSQL = "";
    private System.Drawing.Color _BackColor;

    public event QueryThrownEventHandler QueryThrown;
    public delegate void QueryThrownEventHandler(DataSet dts);
    #endregion

    #region "Propiedades"
    public int Item
    {
        get { return _Item; }
        set { _Item = value; }
    }

    public string Tabla
    {
        get { return _Tabla; }
        set { _Tabla = value; }
    }

    public string Campo
    {
        get { return _Campo; }
        set { _Campo = value; }
    }

    public string AliasCampo
    {
        get { return FieldName; }
        set { FieldName = value; }
    }

    public string ExpresionBd
    {
        get { return _ExpresionBd; }
        set { _ExpresionBd = value; }
    }

    public System.Drawing.Color BackColor
    {
        get { return _BackColor; }
        set { _BackColor = value; }
    }

    public override bool ReadOnly
    {
        get { return base.ReadOnly; }
        set
        {
            base.ReadOnly = value;
            if (value)
            {
                //_BackColor = ColorSel;
            }
        }
    }

    public GridViewDataColumn RelatedColumn
    {
        get { return _relatedColumn; }
        set { _relatedColumn = value; }
    }

    public string Desc_Datafield
    {
        get { return _Desc_Datafield; }
        set { _Desc_Datafield = value; }
    }

    public string Desc_DBTable
    {
        get { return _Desc_DBTable; }
        set { _Desc_DBTable = value; }
    }

    public string Desc_DBPK
    {
        get { return _Desc_DBPK; }
        set { _Desc_DBPK = value; }
    }

    public string Desc_Where
    {
        get { return myWhere; }
        set { myWhere = value; }
    }

    public bool Query_on_Text_Changed
    {
        get { return _queryOnTextChanged; }
        set { _queryOnTextChanged = value; }
    }

    public string ExtraSQL
    {
        get { return _extraSQL; }
        set { _extraSQL = value; }
    }

    public bool Requery
    {
        get { return _requery; }
        set { _requery = value; }
    }
    #endregion


    public string queryDesc(string codigo)
    {
        /*
         * TODO: Describe the quert.
         */
        string functionReturnValue = null;

        /*
        string functionReturnValue = null;
        //lanza la query para resolver la descripción
        _requery = false;
        if (!string.IsNullOrEmpty(codigo))
        {
            ASADB.Connection db = new ASADB.Connection();
            string mySql = "select " + _Desc_Datafield + (!string.IsNullOrEmpty(_extraSQL) ? ", " + _extraSQL : "") + " from " + _Desc_DBTable + " where " + _Desc_DBPK + " = '" + codigo + "'";
            if ((myWhere != null) & !string.IsNullOrEmpty(myWhere))
            {
                mySql = mySql + " and " + myWhere;
            }
            DataSet dts = new DataSet();
            dts = db.fillDts(mySql);
            string tmp = null;
            try
            {
                tmp = dts.Tables(0).Rows(0).Item(0);
            }
            catch
            {
                tmp = "";
            }
            if (!string.IsNullOrEmpty(tmp))
            {
                functionReturnValue = tmp;
            }
            else
            {
                functionReturnValue = "Inexistente.";
            }
            if (QueryThrown != null)
            {
                QueryThrown(dts);
            }
        }
        else
        {
            functionReturnValue = "";
        }
       */
        return functionReturnValue;
    }

}

File: documentation/Karve/CustomControls/DataGrid/Clases/DataGridColumn/Types/DataGridBrowseBoxColumn.vb
Imports Telerik.WinControls.UI
Imports VariablesGlobales

Public Class DataGridBrowseBoxColumn
    Inherits GridViewBrowseColumn

#Region "Variables"
    Protected _Item As Integer
    Protected _Campo As String
    Protected _Tabla As String
    Protected _ExpresionBd As String
    Protected _relatedColumn As GridViewDataColumn
    Protected _BackColor As System.Drawing.Color = Drawing.Color.White

    Private _Desc_Datafield As String
    Private _Desc_DBTable As String 'tabla con la que se rellena el campo descripcion
    Private _Desc_DBPK As String 'referencia de la tabla descripcion (el textbox editable)
    Private myWhere As String 'filtro extra en la consulta
    Private _requery As Boolean = True 'si tiene que volver a tirar la query (se pone en true cuando el texto cambia)
    Private _queryOnTextChanged As Boolean = True 'si tira la query cada vez que el texto cambia o espera al lost focus, si por defecto
    Protected _txtDataWidth As Integer
    Private _extraSQL As String = ""
    Public Event QueryThrown(dts As DataSet)
#End Region

#Region "Propiedades"
    Property Item As Integer
        Get
            Return _Item
        End Get
        Set(value As Integer)
            _Item = value
        End Set
    End Property

    Property Tabla As String
        Get
            Return _Tabla
        End Get
        Set(value As String)
            _Tabla = value
        End Set
    End Property

    Property Campo As String
        Get
            Return _Campo
        End Get
        Set(value As String)
            _Campo = value
        End Set
    End Property

    Property AliasCampo As String
        Get
            Return FieldName
        End Get
        Set(value As String)
            FieldName = value
        End Set
    End Property

    Property ExpresionBd As String
        Get
            Return _ExpresionBd
        End Get
        Set(value As String)
            _ExpresionBd = value
        End Set
    End Property

    Property BackColor As System.Drawing.Color
        Get
            Return _BackColor
        End Get
        Set(value As System.Drawing.Color)
            _BackColor = value
        End Set
    End Property

    Public Overrides Property [ReadOnly] As Boolean
        Get
            Return MyBase.[ReadOnly]
        End Get
        Set(value As Boolean)
            MyBase.[ReadOnly] = value
            If value Then
                _BackColor = ColorSel
            End If
        End Set
    End Property

    Property RelatedColumn As GridViewDataColumn
        Get
            Return _relatedColumn
        End Get
        Set(value As GridViewDataColumn)
            _relatedColumn = value
        End Set
    End Property

    Property Desc_Datafield As String
        Get
            Return _Desc_Datafield
        End Get
        Set(value As String)
            _Desc_Datafield = value
        End Set
    End Property

    Property Desc_DBTable As String
        Get
            Return _Desc_DBTable
        End Get
        Set(value As String)
            _Desc_DBTable = value
        End Set
    End Property

    Property Desc_DBPK As String
        Get
            Return _Desc_DBPK
        End Get
        Set(value As String)
            _Desc_DBPK = value
        End Set
    End Property

    Property Desc_Where As String
        Get
            Return myWhere
        End Get
        Set(value As String)
            myWhere = value
        End Set
    End Property

    Property Query_on_Text_Changed As Boolean
        Get
            Return _queryOnTextChanged
        End Get
        Set(value As Boolean)
            _queryOnTextChanged = value
        End Set
    End Property

    Property ExtraSQL As String
        Get
            Return _extraSQL
        End Get
        Set(value As String)
            _extraSQL = value
        End Set
    End Property

    Property Requery As Boolean
        Get
            Return _requery
        End Get
        Set(value As Boolean)
            _requery = value
        End Set
    End Property
#End Region


    Public Function queryDesc(ByVal codigo As String) As String
        'lanza la query para resolver la descripción
        _requery = False
        If codigo <> "" Then
            Dim db As New ASADB.Connection
            Dim mySql As String = "select " & _Desc_Datafield & IIf(_extraSQL <> "", ", " & _extraSQL, "") & " from " & _Desc_DBTable & " where " & _Desc_DBPK & " = '" & codigo & "'"
            If Not IsNothing(myWhere) And myWhere <> "" Then
                mySql = mySql & " and " & myWhere
            End If
            Dim dts As New DataSet
            dts = db.fillDts(mySql)
            Dim tmp As String
            Try
                tmp = dts.Tables(0).Rows(0).Item(0)
            Catch
                tmp = ""
            End Try
            If tmp <> "" Then
                queryDesc = tmp
            Else
                queryDesc = "Inexistente."
            End If
            RaiseEvent QueryThrown(dts)
        Else
            queryDesc = ""
        End If
    End Function

End Class
